contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,821 @@
+1. If the StatefulSet controller has achieved the 
+[declared target state](#target-state) the StatefulSet controller will 
+[complete any in progress updates](#update-completion).
+1. The controller will [report its status](#status-reporting).
+1. The controller will perform any necessary
+[maintenance of its revision history](#history-maintenance).
+
+### Target State
+The target state of the StatefulSet controller with respect to an individual 
+StatefulSet is defined as follows. 
+
+1. The StatefulSet contains exactly `[0,Spec.Replicas)` Pods.
+1. All Pods in the StatefulSet have the correct 
+[target Pod state](#target-pod-state).
+
+### Target Pod State
+As in the [Controller History](https://github.com/kubernetes/community/pull/594) 
+proposal we define the target Object state of StatefulSetSpec specification type 
+object to be the `.Template` and `.VolumeClaimsTemplate`. The latter is currently 
+immutable, but we will version it as one day this constraint may be lifted. This 
+state provides enough information to generate a Pod and its associated 
+PersistentVolumeClaims. The target Pod State for a Pod in a StatefulSet is as 
+follows.
+1. The Pods PersistentVolumeClaims have been created.
+   - Note that we do not currently delete PersistentVolumeClaims.
+1. If the Pod's ordinal is in the sequence `[0,.Spec.Replicas)` the Pod should 
+have a Ready Condition. This implies the Pod is Running.
+1. If Pod's ordinal is greater than or equal to `.Spec.Replicas`, the Pod 
+should be completely terminated and deleted.
+1. If the StatefulSet's `Spec.UpdateStrategy.Type` is equal to 
+`OnDeleteStatefulSetStrategyType`, no version tracking is performed, Pods 
+can be at an arbitrary version, and they will be recreated from the current 
+`.Spec.Template` and `.Spec.VolumeClaimsTemplate` when the are deleted.
+1. If StatefulSet's `Spec.UpdateStrategy.Type` is equal to 
+`RollingUpdateStatefulSetStrategyType` then the version of the Pod should be 
+as follows.
+    1. If the Pod's ordinal is in the sequence `[0,.Status.CurrentReplicas)`, 
+    the Pod should be consistent with version indicated by `Status.CurrentRevision`.
+    1. If the Pod's ordinal is in the sequence 
+    `[.Status.Replicas - .Status.UpdatedReplicas, .Status.Replicas)`
+    the Pod should be consistent with the version indicated by 
+    `Status.UpdateRevision`.
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`PartitionedStatefulSetStrategyType` then the version of the Pod should be 
+as follows.
+    1. If the Pod's ordinal is in the sequence `[0,.Status.CurrentReplicas)`, 
+    the Pod should be consistent with version indicated by `Status.CurrentRevision`.
+    1. If the Pod's ordinal is in the sequence 
+    `[.Status.Replicas - .Status.UpdatedReplicas, .Status.Replicas)`the Pod 
+    should be consistent with the version indicated by `Status.UpdateRevision`.
+    1. If the Pod does not meet either of the prior two conditions, and if 
```

Comment:
@kow3ns @kargakis the first two lines are exactly the same as rolling update. I am not able to visualize the difference at a given point in time when rolling update is going on and PartitionedRollingUpdate is going on . Do you guys think we need an additional field for the partitioned case ?